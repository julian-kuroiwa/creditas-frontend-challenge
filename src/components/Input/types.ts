import { InputHTMLAttributes } from 'react'

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  label: string
  id: string
  errorMessage?: string
}

export type LabelProps = {
  hasError: boolean
}
