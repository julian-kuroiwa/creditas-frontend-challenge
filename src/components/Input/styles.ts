import styled from 'styled-components'
import { theme } from '~/styles/theme'
import { LabelProps } from './types'

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`

export const Container = styled.div`
  height: 64px;
  position: relative;
  padding: 0px ${theme.spacing.s};
  border-radius: 16px;
  background-color: ${theme.colors.gray.light};
  transition: background-color 0.2s;
  display: flex;
  align-items: center;
  width: 100%;
`
export const Label = styled.label<LabelProps>`
  color: ${({ hasError }) =>
    hasError ? theme.colors.error : theme.colors.gray.medium};
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  line-height: 16px;
  display: block;
  transition: 0.2s;
`

export const Input = styled.input`
  border: none;
  background: transparent;
  outline: none;
  padding: 0px;
  font-size: 16px;
  line-height: 24px;
  color: ${theme.colors.dark};
  margin-top: calc(${theme.spacing.xs} / 2);
  resize: none;
  height: 24px;
  width: 100%;
  z-index: 1;

  ${Container}:has(> &:focus) {
    box-shadow:
      rgba(0, 0, 0, 0.08) 0px 12px 12px,
      rgba(0, 0, 0, 0.08) 0px 0px 8px;
    background-color: ${theme.colors.light};
  }

  &:-webkit-autofill,
  &:not(:placeholder-shown) + ${Label}, &:focus + ${Label} {
    top: 13px;
    font-size: ${theme.font.text.size.xs};
  }
`

export const Error = styled.span`
  font-size: ${theme.font.text.size.xs};
  color: ${theme.colors.error};
  margin-top: ${theme.spacing.xs};
  display: block;
`
