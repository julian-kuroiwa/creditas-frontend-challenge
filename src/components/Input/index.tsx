import * as S from './styles'
import { InputProps } from './types'

const Input = ({ label, id, errorMessage, ...props }: InputProps) => {
  return (
    <>
      <S.Container>
        <S.Input {...props} id={id} />
        <S.Label htmlFor={id} hasError={!!errorMessage}>
          {label}
        </S.Label>
      </S.Container>
      {errorMessage && <S.Error>{errorMessage}</S.Error>}
    </>
  )
}

export default Input
