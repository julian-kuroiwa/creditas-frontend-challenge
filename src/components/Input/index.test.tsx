import { render } from '@testing-library/react'
import Input from '.'

describe('Input', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <Input id="input-test" type="text" label="Input test" />,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('should show an error message if the props errorMessage is passed', () => {
    const errorMessage = 'this is an error message'
    const { getByText } = render(
      <Input
        id="input-test"
        type="text"
        label="Input test"
        errorMessage={errorMessage}
      />,
    )

    expect(getByText(errorMessage)).toBeTruthy()
  })
})
