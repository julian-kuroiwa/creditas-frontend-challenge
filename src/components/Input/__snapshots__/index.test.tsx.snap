// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-FEMpB hHfTmq"
  >
    <div
      class="sc-gjcoXW wmwZ"
    >
      <input
        class="sc-fLDLck djtGcT"
        id="input-test"
        placeholder=" "
        type="text"
      />
      <label
        class="sc-dVBluf epDlHY"
        for="input-test"
      >
        Input test
      </label>
    </div>
  </div>
</DocumentFragment>
`;
